plutus-core/plutus-core/src/PlutusCore/Core/Type.hs

```diff
@@ -67,13 +68,32 @@ data Type tyname uni ann
     | TyApp ann (Type tyname uni ann) (Type tyname uni ann)
     deriving (Show, Functor, Generic, NFData, Hashable)
 
+type BuiltinTag :: GHC.Type -> GHC.Type
+newtype BuiltinTag fun = BuiltinTag
+    { unBuiltinTag :: Int
+    } deriving stock (Show, Generic, Eq, Ord)
+      deriving newtype (NFData, Hashable)
+
+instance (Bounded fun, Enum fun) => Bounded (BuiltinTag fun) where
+    minBound = renum $ minBound @fun
+    maxBound = renum $ maxBound @fun
+
+instance Enum fun => Enum (BuiltinTag fun) where
+    succ     = renum . succ @fun . renum
+    pred     = renum . pred @fun . renum
+    toEnum   = BuiltinTag
+    fromEnum = unBuiltinTag
+
+instance (Pretty fun, Enum fun) => Pretty (BuiltinTag fun) where
+    pretty = pretty @fun . renum
+
 data Term tyname name uni fun ann
     = Var ann name -- ^ a named variable
     | TyAbs ann tyname (Kind ann) (Term tyname name uni fun ann)
     | LamAbs ann name (Type tyname uni ann) (Term tyname name uni fun ann)
     | Apply ann (Term tyname name uni fun ann) (Term tyname name uni fun ann)
     | Constant ann (Some (ValueOf uni)) -- ^ a constant term
-    | Builtin ann fun
+    | Builtin ann (BuiltinTag fun)
```

Comment:
UNPACK? here and in UPLC

Reply:
That [happens by default](https://downloads.haskell.org/%7Eghc/latest/docs/html/users_guide/using-optimisation.html#ghc-flag--funbox-small-strict-fields), but ok.

Reply:
But I did forget to add a bang :facepalm:

Reply:
Ah, well, we don't have bangs in TPLC and it's there in UPLC.

Reply:
Yeah, but I think it's good to be explicit, and I think you'll get an error if it doesn't fire? So maybe just in UPLC then.